src/armature/Generator.ts

```diff
@@ -139,8 +156,12 @@ export class GeneratorInstance {
     public growIfPossible() {
         const originalLength = this.model.nodes.length;
 
+        let choiceProbability = 1;
+
         while (this.model.nodes.length === originalLength && this.spawnPoints.length > 0) {
-            // Remove a random spawn point from the list of active points
+            // Remove a random spawn point from the list of active points, updating the probability
+            // given the number of choices we have
+            choiceProbability /= this.spawnPoints.length;
```

Comment:
Why is this updated on every loop iteration? Suppose in one iteration, the `generator` call produces no new nodes and the loop runs again, and in this iteration it produces some geometry. Then `choiceProbability = (this.spawnPoints.length * (this.spawnPoints.length - 1))^-1`. It's not clear to me what significance this quantity has.

Reply:
So it's supposed to represent the probability of arriving at the specific sample, so if you pick a spawn point to advance out of n options, it's saying the model you arrive at had a 1/n chance of being picked. The loop thing is a little odd because not every choice in the grammar adds geometry, so it's sort of like, one "expansion" ends up making more than one choice.

Thinking about it some more though, since some of the available spawn points are still available as options to expand. I'm trying to read through the appendix of <a href="https://cocolab.stanford.edu/papers/RitchieEtAl2015-SIGGRAPH.pdf">SOSMC</a> again to try to get some more insight since it goes into this

Reply:
Ok so the line in section 4 of SOSMC that I'm trying to represent is this:

*p*<sub>*n*</sub>(**x**<sub>*n*</sub>) = *p*<sub>*n*−1</sub>(**x**<sub>*n*−1</sub>) · *p*(**x**<sub>*n*</sub>|**x**<sub>*n*−1</sub>)

*p*<sub>*n*−1</sub>(**x**<sub>*n*−1</sub>) is the parent probability, which is `this.probability` at this point. The paper defines *p*(**x**<sub>*n*</sub>|**x**<sub>*n*−1</sub>) as:

𝚷<sub>􏰃*i*=1</sub><sup>|**x**<sub>*n*</sub> \\ **x**<sub>*n*−1</sub>|</sup> *p*(*x*<sub>*n*, *i*</sub> | *x*<sub>*n*,1:(*i*−1)</sub>,*x*<sub>*n*−1</sub>)

and then with the stochastic choice policy, they define it as:

𝚷<sub>􏰃*i*=1</sub><sup>|**x**<sub>*n*</sub> \\ **x**<sub>*n*−1</sub>|</sup> 1/(number of subcomputations that can be continued at this point)

So right now, we are multiplying by 1/(number of subcomputations that can be continued at this point). Although I'm still not entirely sure what is encompassed by the range from 1 to |**x**<sub>*n*</sub> \\ **x**<sub>*n*−1</sub>|, since I thought by definition there was only one more element in **x**<sub>*n*</sub> than **x**<sub>*n*−1</sub>, making the product over the range sort of irrelevant.

Reply:
Although they do mention each item in **x**<sub>*n*</sub> adds a piece of geometry. I guess in our case we shouldn't keep multiplying, and only divide by the number of available spawn points when we've picked one that will for sure add geometry?

Reply:
I suppose that would technically be correct, but that would be expensive to compute. We can approximate it by only setting `choiceProbability` to `1/numSpawnPoints` once we actually create more geometry (this would also handle the case where no generator produces geometry, and `choiceProbability` should be 1).

Reply:
So upon further reading, I think 𝚷<sub>􏰃*i*=1</sub><sup>|**x**<sub>*n*</sub> \\ **x**<sub>*n*−1</sub>|</sup> means "for each random choice made", since **x**<sub>n</sub> refers to the *set of random choices* that leads to producing *n* pieces of geometry, not the *n* pieces themselves. So every time we make a random choice, we multiply in the probability of making that choice.

In that interpretation, I think maybe the current implementation might actually be ok. Thoughts?

Reply:
Makes sense!